codec-http/src/main/java/io/netty5/handler/codec/http/headers/ReadOnlyHttpHeaders.java

```diff
@@ -0,0 +1,626 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.netty5.handler.codec.http.headers;
+
+import io.netty5.handler.codec.http.headers.HeaderUtils.CookiesByNameIterator;
+import io.netty5.util.AsciiString;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static io.netty5.handler.codec.http.HttpHeaderNames.COOKIE;
+import static io.netty5.handler.codec.http.HttpHeaderNames.SET_COOKIE;
+import static io.netty5.handler.codec.http.headers.DefaultHttpSetCookie.parseSetCookie;
+import static io.netty5.handler.codec.http.headers.HeaderUtils.DEFAULT_HEADER_FILTER;
+import static io.netty5.handler.codec.http.headers.HeaderUtils.domainMatches;
+import static io.netty5.handler.codec.http.headers.HeaderUtils.parseCookiePair;
+import static io.netty5.handler.codec.http.headers.HeaderUtils.pathMatches;
+import static io.netty5.util.AsciiString.contentEquals;
+import static io.netty5.util.AsciiString.contentEqualsIgnoreCase;
+import static java.util.Collections.emptyIterator;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+final class ReadOnlyHttpHeaders implements HttpHeaders {
+    private final CharSequence[] keyValuePairs;
+    private final boolean validateCookies;
+
+    ReadOnlyHttpHeaders(final CharSequence... keyValuePairs) {
+        this(true, keyValuePairs);
+    }
+
+    ReadOnlyHttpHeaders(boolean validateCookies, final CharSequence... keyValuePairs) {
+        if ((keyValuePairs.length & 1) != 0) {
+            throw new IllegalArgumentException("keyValuePairs length must be even but was: " + keyValuePairs.length);
+        }
+        this.keyValuePairs = requireNonNull(keyValuePairs);
+        this.validateCookies = validateCookies;
+    }
+
+    private static int hashCode(final CharSequence name) {
+        return AsciiString.hashCode(name);
+    }
+
+    private static boolean equals(final CharSequence name1, final CharSequence name2) {
+        return contentEqualsIgnoreCase(name1, name2);
+    }
+
+    private static boolean equalsValues(final CharSequence name1, final CharSequence name2) {
+        return contentEquals(name1, name2);
+    }
+
+    @Nullable
+    @Override
+    public CharSequence get(final CharSequence name) {
+        final int nameHash = hashCode(name);
+        final int end = keyValuePairs.length - 1;
+        for (int i = 0; i < end; i += 2) {
+            final CharSequence currentName = keyValuePairs[i];
+            if (nameHash == hashCode(currentName) && equals(currentName, name)) {
+                return keyValuePairs[i + 1];
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public CharSequence getAndRemove(final CharSequence name) {
+        throw new UnsupportedOperationException();
```

Comment:
I think this implementation violate the API contract. `getAndRemove` etc are not marked as "optional" methods.

Reply:
So we need to mark them as optional… or would you rather drop the read-only implementation?

The existing read-only implementation of the old API does not appear to be used by Netty itself.

Reply:
I am not sure... I guess maybe remove it ?

Reply:
I'm partial to removing the read-only implementation since the motivation appears to be iteration performance. However, if we move to a B-tree based multi-map implementation, the iteration performance will probably get pretty close to the flat array for most collections of headers seen in practice.